namer/namer.cc

```diff
@@ -15,16 +18,1336 @@
+    // Contains all type members defined in the file.
+    vector<FoundTypeMember> _typeMembers;
+    // Contains all symbol references in the file.
+    vector<core::SymbolRef> _symbols;
+    // Contains all method and class modifiers (e.g. private/public/protected).
+    vector<Modifier> _modifiers;
+
+    FoundNameRef addDefinition(FoundNameRef ref) {
+        _definitions.push_back(ref);
+        return ref;
+    }
+
+public:
+    FoundNames() = default;
+    FoundNames(FoundNames &&names) = default;
+    FoundNames(const FoundNames &names) = delete;
+    ~FoundNames() = default;
+
+    FoundNameRef addClass(FoundClass &&klass) {
+        const u4 idx = _klasses.size();
+        _klasses.push_back(move(klass));
+        return addDefinition(FoundNameRef(NameKind::Class, idx));
+    }
+
+    FoundNameRef addClassRef(FoundClassRef &&klassRef) {
+        const u4 idx = _klassRefs.size();
+        _klassRefs.push_back(move(klassRef));
+        return FoundNameRef(NameKind::ClassRef, idx);
+    }
+
+    FoundNameRef addMethod(FoundMethod &&method) {
+        const u4 idx = _methods.size();
+        _methods.push_back(move(method));
+        return addDefinition(FoundNameRef(NameKind::Method, idx));
+    }
+
+    FoundNameRef addStaticField(FoundStaticField &&staticField) {
+        const u4 idx = _staticFields.size();
+        _staticFields.push_back(move(staticField));
+        return addDefinition(FoundNameRef(NameKind::StaticField, idx));
+    }
+
+    FoundNameRef addTypeMember(FoundTypeMember &&typeMember) {
+        const u4 idx = _typeMembers.size();
+        _typeMembers.push_back(move(typeMember));
+        return addDefinition(FoundNameRef(NameKind::TypeMember, idx));
+    }
+
+    FoundNameRef addSymbol(core::SymbolRef symbol) {
+        const u4 idx = _symbols.size();
+        _symbols.push_back(move(symbol));
+        return FoundNameRef(NameKind::Symbol, idx);
+    }
+
+    void addModifier(Modifier &&mod) {
+        _modifiers.push_back(move(mod));
+    }
+
+    const vector<FoundNameRef> &definitions() const {
+        return _definitions;
+    }
+
+    const vector<FoundClass> &klasses() const {
+        return _klasses;
+    }
+
+    const vector<FoundClassRef> &klassRefs() const {
+        return _klassRefs;
+    }
+
+    const vector<FoundMethod> &methods() const {
+        return _methods;
+    }
+
+    const vector<FoundStaticField> &staticFields() const {
+        return _staticFields;
+    }
+
+    const vector<FoundTypeMember> &typeMembers() const {
+        return _typeMembers;
+    }
+
+    const vector<Modifier> &modifiers() const {
+        return _modifiers;
+    }
+
+    const vector<core::SymbolRef> &symbols() const {
+        return _symbols;
+    }
+
+    friend FoundNameRef;
+};
+
+FoundClassRef &FoundNameRef::klassRef(FoundNames &foundNames) {
+    ENFORCE(kind() == NameKind::ClassRef);
+    ENFORCE(foundNames._klassRefs.size() > idx());
+    return foundNames._klassRefs[idx()];
+}
```

Comment:
It seems like there's a lot of boilerplate and indirection here.

Can you walk me through a little bit of why this boilerplate and indirection pays off down the line?

Reply:
I'm open to defenses here, but it looks to me like these helpers might not be pulling their weight:

- We only ever iterate over `foundNames->definitions()` and `foundNames->modifiers()`. The sub-types (things like methods, classes, etc, don't benefit from being colocated in one contiguous array.
- The helpers we have like `name.symbols()` are already "not type safe" in that they ENFORCE that they're used correctly, which is not much better than `cast`'ing at runtime.
- There is a lot of boilerplate, which makes it hard at a glance to tell what's going on (i.e., which are the important functions to read, vs which are just there to make it all fit).
- It doesn't use names / idioms consistent with the rest of the codebase (i.e., `SymbolRef` has a `.exists()` and `.data()` method, as do `NameRef` and `FileRef`, so when I see `FooRef` i can immediately guess at what methods it might have.

Do we benefit from having all these separate data structures? When these definitions make it into the symbol table eventually, they're all going to get squashed into one data structure anyways (`Symbol` / `SymbolRef`).

Reply:
> The sub-types (things like methods, classes, etc, don't benefit from being colocated in one contiguous array.

I initially did not have the contiguous array, and it dirtied the diff considerably for test exp files (it reorders symbol tables). I also was not sure if it could cause correctness issues, e.g. trying to consume a reference to a symbol before it's defined. For example, can a method ever be the `ctx.owner` of a class?

I can see if separating them again causes issues.

> It doesn't use names / idioms consistent with the rest of the codebase (i.e., SymbolRef has a .exists() and .data() method, 

I consciously chose to _not_ have a `data()` method as there's no parent struct to return (intentionally so, as inheritance isn't needed here). There is an `exists()`.

I need some way to generically refer to a name that may be already defined (symbolref) or that needs to be defined (method/class/etc). 

If I am able to move away from the single array, I might be able to pare it down more -- only methods and classes can own things, for example. That would cut down on the boilerplate.

Reply:
@jez see update above, Having a single array is a correctness thing here to match old behaviour.

Do you still want it done differently knowing that?